marklogic-client-api/src/main/java/com/marklogic/client/impl/OkHttpServices.java

```diff
@@ -238,45 +237,138 @@ private FailedRequest extractErrorFields(Response response) {
-        @Override
-        public boolean verify(String hostname, SSLSession session) {
-          return true;
-        }
-      };
-    } else if (verifier == SSLHostnameVerifier.COMMON) {
-      hostnameVerifier = null;
-    } else if (verifier == SSLHostnameVerifier.STRICT) {
-      hostnameVerifier = null;
-    } else if (verifier != null) {
-      hostnameVerifier = new SSLHostnameVerifier.HostnameVerifierAdapter(verifier);
-    }// else {
-    //  throw new IllegalArgumentException(
-    //    "Null SSLContext but non-null SSLHostnameVerifier for client");
-    //}
-    connect(host, port, database, user, password, kerberosOptions, authenType, sslContext, trustManager, hostnameVerifier);
-  }
-
-  private void connect(String host, int port, String database, String user, String password, Map<String,String> kerberosOptions,
-                       Authentication authenType, SSLContext sslContext, X509TrustManager trustManager,
-                       HostnameVerifier verifier) {
-    logger.debug("Connecting to {} at {} as {}", new Object[]{host, port, user});
-
-    if (host == null) throw new IllegalArgumentException("No host provided");
-
+  public void connect(String host, int port, String database, SecurityContext securityContext){
+	String user = null;
+	Map<String, String> kerberosOptions = null;
+	String password = null;
+	Authentication type = null;
+	SSLContext sslContext = null;
+	SSLHostnameVerifier sslVerifier = null;
+	X509TrustManager trustManager = null;
+	String authorizationTokenValue = null;
+    Credentials credentials = null;
+    final Map<String,CachingAuthenticator> authCache = new ConcurrentHashMap<String,CachingAuthenticator>();
+    CachingAuthenticator authenticator = null;
+    Interceptor interceptor = null;
+    
+    if (host == null) 
+    	throw new IllegalArgumentException("No host provided");
+    
+	if (securityContext instanceof BasicAuthContext || ((BasicAuthContext)securityContext).getSSLContext()!=null) {
+		BasicAuthContext basicContext = (BasicAuthContext) securityContext;
+		user = basicContext.getUser();
+		password = basicContext.getPassword();
+		if (user == null) 
+			throw new IllegalArgumentException("No user provided");
+	    if (password == null) 
+	    	throw new IllegalArgumentException("No password provided");
+		credentials = new Credentials(user, password);
+		logger.debug("Connecting to {} at {} as {} using Basic Authentication.", new Object[]{host, port, user});
+		type = Authentication.BASIC;
+		if (basicContext.getSSLContext() != null) {
+			sslContext = basicContext.getSSLContext();
+			if (basicContext.getTrustManager() != null)
+				trustManager = basicContext.getTrustManager();
+			if (basicContext.getSSLHostnameVerifier() != null) {
+				sslVerifier = basicContext.getSSLHostnameVerifier();
+			} else {
+				sslVerifier = SSLHostnameVerifier.COMMON;
+			}
+			interceptor = new HTTPBasicAuthInterceptor(credentials);
+	        checkFirstRequest = false;
+		}
```

Comment:
If I recall correctly, an earlier version moved the SSL setup out of the branches and moved the other code from the branches into methods called from the branch by casting the security context and passing the client builder. 

Would that approach still work?

Reply:
Included configureAuthentication overloads. Sorry it was missed during the branch change.

Reply:
Would it work to pass the SecurityContext and OkHttpClient.Builder to each configureAuthentication() method?

As in

    if (securityContext instanceof BasicAuthContext) {
        clientBldr = configureAuthentication((BasicAuthContext) securityContext, clientBldr);
    } else  (securityContext instanceof DigestAuthContext) {
        clientBldr = configureAuthentication((DigestAuthContext) securityContext, clientBldr);
    ... other tests and casts on calls ...

In other words, to move the code that accesses the security context into the configureAuthentication() methods?

Also, given that the SSL getters are defined on SecurityContext, could the SSL configuration be performed once outside of the branches on the specific instances of SecurityContext?

Is that possible?